### ISSUE TITLE
2021.09 Failed update / Startup crash on kobo aura one

### ISSUE THREAD

**Issue by Thra11** [koreader/koreader#8251]
* KOReader version: 2021.09
* Device: Kobo Aura One

#### Issue
Attempted OTA update. Proceeded to download as normal. When asked, pressed button to install now. Progress bar filled up, then message appeared saying something like, "Update failed". Afterwards, any attempt to launcher koreader resulted in an immediate crash.

I tried manually downloading the release and extracting it onto the device vis USB, in case the OTA update had corrupted some files. However, it still crashed.

I was able to manually downgrade it to 2021.08 by copying the release over USB. 2021.08 was working before and appears to be working fine once more.

#### Steps to reproduce
1. Perform OTA update to 2021.09 (on a Kobo Aura One, from 2021.08)

##### `crash.log` (if applicable)

<details>
  <summary>crash.log</summary>

```
---------------------------------------------
                launching...
  _  _____  ____                _
 | |/ / _ \|  _ \ ___  __ _  __| | ___ _ __
 | ' / | | | |_) / _ \/ _` |/ _` |/ _ \ '__|
 | . \ |_| |  _ <  __/ (_| | (_| |  __/ |
 |_|\_\___/|_| \_\___|\__,_|\__,_|\___|_|

 It's a scroll... It's a codex... It's KOReader!

 [*] Current time: 09/22/21-23:43:34
 [*] Version: v2021.09

ffi.load: blitbuffer
09/22/21-23:43:34 INFO  initializing for device Kobo_daylight
09/22/21-23:43:34 INFO  framebuffer resolution: {
    ["h"] = 1872,
    ["w"] = 1404,
}
[ko-input] Forked off fake event generator (pid: 1945)
ffi.load: libs/libmupdf.so
Bus error
!!!!
Uh oh, something went awry... (Crash n°1: 09/22/21 @ 23:43:35)
Running FW 4.19.14123 on Linux 3.0.35+ (#5030 PREEMPT Wed Oct 25 10:25:24 CST 2017)
Attempting to restart KOReader . . .
!!!!
```
</details>

Longer [crash.log](https://github.com/koreader/koreader/files/7219747/crash.log)

**Comment by Thra11** [comment]
Think it might switch from the crash I quoted above to this crash at the point where I manually extracted the 2021.09 update over USB:
<details>
<summary>crash.log</summary>
```
---------------------------------------------
                launching...
  _  _____  ____                _
 | |/ / _ \|  _ \ ___  __ _  __| | ___ _ __
 | ' / | | | |_) / _ \/ _` |/ _` |/ _ \ '__|
 | . \ |_| |  _ <  __/ (_| | (_| |  __/ |
 |_|\_\___/|_| \_\___|\__,_|\__,_|\___|_|

 It's a scroll... It's a codex... It's KOReader!

 [*] Current time: 09/23/21-16:35:21
 [*] Version: v2021.09

ffi.load: blitbuffer
09/23/21-16:35:21 INFO  initializing for device Kobo_daylight
09/23/21-16:35:21 INFO  framebuffer resolution: {
    ["w"] = 1404,
    ["h"] = 1872,
}
[ko-input] Forked off fake event generator (pid: 1305)
ffi.load: libs/libmupdf.so
ffi.load (warning): /mnt/onboard/.adds/koreader/libs/libz.so.1: file too short
./luajit: ./setupkoenv.lua:27: Not able to load dynamic library: libs/libmupdf.so
stack traceback:
    [C]: in function 'error'
    ./setupkoenv.lua:27: in function 'load'
    ./ffi/mupdf.lua:25: in main chunk
    [C]: in function 'require'
    frontend/document/canvascontext.lua:14: in main chunk
    [C]: in function 'require'
    ./reader.lua:176: in main chunk
    [C]: at 0x00014119
!!!!
Uh oh, something went awry... (Crash n°4: 09/23/21 @ 16:35:21)
Running FW 4.19.14123 on Linux 3.0.35+ (#5030 PREEMPT Wed Oct 25 10:25:24 CST 2017)
Attempting to restart KOReader . . .
!!!!
```
</details>

**Comment by Frenzie** [comment]
Could there have been a lack of free space?

**Comment by Thra11** [comment]
I checked after the OTA update and there was ~86MB free, so that may have affected the update.

However, I deleted some files to free up another ~300MB before manually copying 2021.09 over USB, so I doubt that was still an issue when it kept crashing afterwards.

**Comment by Frenzie** [comment]
File too short sounds like you should at the very least overwrite that file some more with the original. ;-) But I'd just do the whole thing again so you're sure there aren't more of them.

**Comment by NiLuJe** [comment]
Yep, both a SIGBUS and a file too short point at storage space constraints breaking the update ;).

(FWIW, you need roughly two to three times the size of the full folder for an OTA update, so, 300MB is already possibly on the dangerous side of things).

If you're really hurting for space that much, you'll have to handle updates manually.

**Comment by Thra11** [comment]
>If you're really hurting for space that much, you'll have to handle updates manually.

The latter 2021.09 update _was_ manual and still wouldn't run, so I still suspect there might be something wrong with the update.

I didn't wipe the existing koreader directory before updating (Think there are some cache and settings files mixed in with the firmware, which I didn't want to lose), so maybe there are files there that aren't part of the 2021.09 update but still affect it in some way? Seems odd that 2021.08 isn't be affected if that was the case.

**Comment by Thra11** [comment]
Also, when I diff'd the device's koreader directory against the update, I noticed there were a lot of files in there that appear to have accumulated from older releases. For example, the device has both `libcrypto.so.1.0.0` and `libcrypto.so.1.1`, while the update only contains `libcrypto.so.1.1`. Does koreader not clean up old files that it doesn't overwrite with a newer version?

**Comment by NiLuJe** [comment]
> Does koreader not clean up old files that it doesn't need in the newer version?

Not currently, although we could do that in onetimeupdate. But its irrelevant because they're safely ignored anyway ;).

----

The loader error is pretty straightforward: the file is corrupt. It's the *right* file, it's just, well, corrupt ;o). Running out of space would certainly explain it for an OTA.
For a manual update, too, but you should have caught that ;).

That leaves storage failure?

**Comment by NiLuJe** [comment]
A broken update file is highly unlikely, as would be an incomplete one. (Nothing's changed on that front in a while, and the OTA does an integrity check).

**Comment by Thra11** [comment]
> (FWIW, you need roughly two to three times the size of the full folder for an OTA update, so, 300MB is already possibly on the dangerous side of things).

When starting an OTA update, it might be useful to check the free space and warn the user if it's below a "safe threshold". I'm assuming that the exact size of the update might vary slightly (It's 86MB for the last two kobo updates) and might not be known before downloading it, but a carefully chosen hard-coded figure would probably do the job. I'm assuming that the consequences are likely to be similar to what I experienced, so a shouty confirmation message along the lines of, "You don't have enough space to update safely. This could break KOReader", would be helpful.

> The loader error is pretty straightforward: the file is corrupt. It's the right file, it's just, well, corrupt ;o). Running out of space would certainly explain it for an OTA.

I should have checked the md5sum of libz.so.1 while it was in the broken state, to compare it to the corresponding file in the update. Unfortunately, I didn't, so it was probably truncated but I won't be able to confirm it.

> That leaves storage failure?

Always a possibility. The device is getting old, and I think storage can get finicky when low on space (and it's vfat, so not the most robust filesystem).

> But it's irrelevant because they're safely ignored anyway ;).

Is that strictly true? Unlikely to be relevant to this issue, but don't plugins get loaded simply by existing in `plugins` and having files with the right names?

**Comment by Thra11** [comment]
BTW, not planning to do anything else risky tonight, but I'll try updating from 2021.08 to 2021.09 again tomorrow, both manually and using OTA, and see what happens.

**Comment by NiLuJe** [comment]
> 
> Is that strictly true? Unlikely to be relevant to this issue, but don't plugins get loaded simply by existing in `plugins` and having files with the right names?

Removed plugins are flagged as "do not load" in the plugin loader to avoid this ;).

**Comment by NiLuJe** [comment]
> Assuming that the consequences are likely to be similar to what I experienced, a shouty confirmation message along the lines of, "You don't have enough space to update safely. This could break KOReader", would be helpful.

I've thought about it in the past, but it's mildly annoying to:
* Devise a "good enough" sweet spot value
* Actually do the check in a platform-specific manner

(Not impossible, just a lot of boring code)

And, in the end, that's a fairly rare edge-case which usually ends up with a very explicit failure ;p. (I mean, we've probably seen something like two reports of this in the last five years).

TL;DR: Low priority ;).
